pytorch_lightning/utilities/argparse.py

```diff
@@ -19,11 +19,13 @@
 from argparse import _ArgumentGroup, ArgumentParser, Namespace
 from contextlib import suppress
 from functools import wraps
-from typing import Any, Callable, Dict, List, Tuple, Type, Union
+from typing import Any, Callable, cast, Dict, List, Tuple, Type, TypeVar, Union
 
 import pytorch_lightning as pl
 from pytorch_lightning.utilities.parsing import str_to_bool, str_to_bool_or_int, str_to_bool_or_str
 
+_T = TypeVar("_T", bound=Callable[..., Any])
```

Comment:
Is the Callable parametrization helpful?

Why not include the input arguments?

Is the following equal?
```suggestion
_T = TypeVar("_T", bound=Callable)
```

Reply:
> Is the Callable parametrization helpful?

Yes, it enables type-checkers to detect bad usages of the decorator. E.g.

```python
_defaults_from_env_vars(None)  # static-checker: error `None` not callable
```

> Why not include the input arguments?

Because `_defaults_from_env_vars` is agnostic to the actual arguments of the function that it is decorating: it ultimately accepts `*args, **kwargs` and passes them through. Thus attempting to include input annotations would make `_T` overly narrow.
 
> Is the following equal? `Callable[..., Any]` vs `Callable`

Yes, those are equivalent.

Reply:
@carmocca just a minor note about something I came across this past week:

pyright (microsoft's type checker) doesn't treat `Callable[..., Any]` and `Callable` as being equivalent when running in "strict" mode. The latter is treated as `Callable[unknown, unknown]`.